sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala

```diff
@@ -432,18 +432,34 @@ object ResolveDefaultColumns extends QueryErrorsBase
       targetType: DataType,
       colName: String): Option[Expression] = {
     expr match {
-      case l: Literal if !Seq(targetType, l.dataType).exists(_ match {
+      case l: Literal => defaultValueFromWiderType(l, targetType, colName)
```

Comment:
To explain this, the analyze() function in this class for strings does the following

1. analyze/ checkAnalyze
2. constant fold
3. cast

However, because we add the rule in TypeCoercion, it runs before constant folding.  Hence, relaxing the cast code to also take Expression instead of Literal.